### ISSUE TITLE
[RFC] Native GPU-Direct RDMA implementation of collective ops

### ISSUE THREAD

**Issue by byronyi** [tensorflow/tensorflow#18232]
In a [recent CL](a0e0685ca974e484de9200caf8c414dcb55277bb), the remote memory access based collective ops (`CollectiveRemoteAccess`) and a local implementation (`CollectiveRemoteAccessLocal`) has been added (looks like @poxvoculi's glorious work).

I am wondering if there would be enough space for a verbs implementation of the remote access collective ops, or its design has been carried out and TF team will release its implementation soon? If there is no such plan in a short term, I would love to extend the GDR runtime to the collective ops. In particular, I would like to request for comments to the future of the GDR runtime, as it seems the contrib modules may eventually be moved out to other places (https://github.com/tensorflow/tensorflow/pull/16984#issuecomment-365918193).

Any suggestion from the TF team?

**Comment by poxvoculi** [comment]
I expected you would be interested in this.  It's going to take some more time, probably a few weeks, in order to get all of the pieces checked in (it's a large amount of new code, altogether).  At that point I'd be happy to discuss with you how to extend the GDR runtime to support collective ops.  The design goal is very much to allow for high-performance extensions since it's impossible to offer one implementation that is optimal for all circumstances.  Towards that end it will be very useful to have feedback from users adapting TF to different platforms.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by byronyi** [comment]
Thanks for further explanation (and also answering questions in mailing list 😆). We are still testing the distribution strategy API and its GDR extension, and hopefully get it upstream before v2.0. 

Looking further, as DistributionStrategy will be a prominent API for scaling model training, we hope that the collective ops could be extended by networking plugin (after contrib is removed), i.e. RecvBuf, along with the RecvTensor RPC. We could discuss more details later.

**Comment by yanivbl6** [comment]
hi @poxvoculi , @byronyi 
I would like to have the ability to initialize an implementation dependent context for a collective, preferably during the construction of the CollectiveReduceOpKernel.
As it stands, the collective kernel seem oblivious to the collective implementation/executor at construction time (it is also unaware of the network- which I would like to be).
The obvious work-around for me is to wait to the first call to computeAsync (or ExectureAsync, but I would still need to pass context to it) and add an initialization over there, but I would rather not go there.

Do you think it is possible to expose the collective_executor earlier on?

**Comment by byronyi** [comment]
@yanivbl6 I did not quite get what you are trying to accomplish. Could you be more specific here? What information you are trying to expose to the OpKernel through the context?

**Comment by yanivbl6** [comment]
The implementation I am working on rely on each collective operator having it's own unique context- which would be eventually bounded to collective-specific hardware resources.

To initialize this context, I would need:
1. The collective_executor (Otherwise I can not do any implementation-specific initialization- right?)
2. Remote access- some remote access capability with the other workers in the collective, to coordinate prior to the usage of the operators.

I have this information during kernel execution, but I require it during init.
I am trying writing some draft to present the API changes I had in mind.

**Comment by byronyi** [comment]
I see. Are we talking about collective ops offloaded to NIC/switch? This would be very interesting.

I do not know how to use such an API though, but if it looks like NCCL, you could take a look at 44f117cecbdcea196ef7ccf93c01c6425ad3232a, which implements multi-node collectives by graph rewrite. I guess similarly you could replace standard collective ops with your own implementation. My hunch is that you do not need most functionalities in CollectiveExecutor as the reduce/broadcast is handled by the hardware anyway. 

But I could be wrong :)

**Comment by poxvoculi** [comment]
@yanivbl6 you can completely replace the collective implementation by providing an alternate implementation of CollectiveExecutorMgrInterface (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/collective.h#L203)
at service creation, via this function:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/distributed_runtime/rpc/grpc_server_lib.h#L89
You can reuse/extend any part of the current implementation via C++ inheritance, or completely bypass it with an alternative.  If you haven't already, look at how ops are initialized here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/collective_ops.cc

In the default implementation the appropriate collective executor can be gotten from the OpKernelContext via 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/op_kernel.h#L1030

Outside of an Op execution (e.g. when handling an incoming RPC) you can find the correct CollectiveExecutor via WorkerEnv::collective_executor_mgr->FindOrCreate();

**Comment by yanivbl6** [comment]
@byronyi 
> I see. Are we talking about collective ops offloaded to NIC/switch? This would be very interesting.
Yes, that's the general idea. I can gladly share more if you are interested.

The API I am currently checking is similar to Facebook's gloo, a.k.a , persistent collectives. 
So it's somehow different to what is being used by nccl and MPI (will be added to MPI4.0). It has strict limitation on parameters (determined during initialization) but allows more flexible ordering in runtime.

@poxvoculi 
My understanding was the intended way to add an implementation is to add a new collective-executor that utilizes new collective implementations,  but still rely on the same Collective operator, with its  CollectiveOpKernel -is that correct?

For now, I think using  and new kernel that inherits from CollectiveOpKernel would work.

**Comment by poxvoculi** [comment]
@yanivbl6 : 

> My understanding was the intended way to add an implementation is to add a new collective-executor that utilizes new collective implementations, but still rely on the same Collective operator, with its CollectiveOpKernel -is that correct?

Yes
